Evolutionary_loop/Antenna_Performance_Metric/XFintoARA.py

```diff
@@ -0,0 +1,151 @@
+import os           # For changing the permissions
+
+
+### GLOBALS ###
+# CORRECTED  Frequency List in MHz
+freqVals = [ 83.33, 100.00, 116.67, 133.33, 150.00, 166.67, 183.34, 200.00,
+            216.67, 233.34, 250.00, 266.67, 283.34, 300.00, 316.67, 333.34, 
+            350.00, 366.67, 383.34, 400.01, 416.67, 433.34, 450.01, 466.67, 
+            483.34, 500.01, 516.68, 533.34, 550.01, 566.68, 583.34, 600.01, 
+            616.68, 633.34, 650.01, 666.68, 683.35, 700.01, 716.68, 733.35, 
+            750.01, 766.68, 783.35, 800.01, 816.68, 833.35, 850.02, 866.68,
+            883.35, 900.02, 916.68, 933.35, 950.02, 966.68, 983.35, 1000.00,
+            1016.70, 1033.40, 1050.00, 1066.70]
+numFreq = 60
+
+# Strings used in the .dat file
+head1_a = "freq : "
+head1_b = " MHz"
+head2   = "SWR : 1.965000"
+head3   = " Theta     Phi     Gain(dB)     Gain     Phase(deg) "
+
+# .uan file location
+# uanLoc = '/home/suren/Desktop/OSU Research/XFintoARAmod/' 
+# location on machtay1 "/users/PAS0654/machtay1/BiconeEvolution/current_antenna_evo_build/"\
+#                      "XF_Loop/Evolutionary_Loop/Antenna_Performance_Metric/"
+uanLoc = "/fs/ess/PAS1960/BiconeEvolutionOSC/BiconeEvolution"\
+         "current_antenna_evo_build/XF_Loop/Evolutionary_Loop/"\
+         "Antenna_Performance_Metric/"
+
+# .dat file saving location
+# datLoc = '/home/suren/Desktop/OSU Research/XFintoARAmod/'
+datLoc = uanLoc # currently, we are saving files in the same directory
+
+
+### DEFINITIONS ###
+
+# Reads the (indiv)freqNum.uan file
+def readFile(indiv, freqNum):
+  n = 37
+  m = 73
+  # uanName = uanLoc+str(indiv)+"_"+str(freqNum)+".uan"
+  # changed from g.indiv + 1 to just indiv 
+
+  # The following (commented out) uanName is the only part that XFintoARA2.py
+  # differs from this script.
+  # uanName = \
+  # f'{g.WorkingDir}/Run_Outputs/{g.RunName}/uan_files/{g.gen}_uan_files/{g.gen}_{indiv}_{freqNum}.uan'
+  uanName = \
+  f'{g.WorkingDir}/Run_Outputs/{g.RunName}/uan_files/{g.gen}_{indiv}_{freqNum}.uan'
+  f = open(uanName, "r")
+
+  for a in range(18):
+    f.readline()
+
+  # ie. m rows n cols // Jason 0612,23
+  mat = [["0" for x in range(n)] for y in range(m)]
+  for i in range(n):
+    for j in range(m):
+
+      line = f.readline()
+      lineList = line.split(" ");
+      lineList[0] = lineList[0] + " " + "\t" + " "
+      lineList[1] = lineList[1] + " " + "\t" + " "
+      linearGainNum = float(lineList[2])
+      # %'s are for interpolation (2 digits after decimal point)
+      linearGainNum = "%.2f" % 10 ** (linearGainNum / 10)
+      linearGain = str(linearGainNum) + "     " + "\t" + "    "
+      lineList[2] = "%.2f" % float(lineList[2]) + "     " + "\t" + "   "
+      lineList[4] = "%.2f" % float(lineList[5])
+      lineFinal = lineList[0] + lineList[1] + \
+                  lineList[2] + linearGain + lineList[4]+"\n"
+  
+      mat[j][i] = lineFinal
+  f.close()        
```

Comment:
This can be removed if you use the `with` statement.

Reply:
Is it still better to include `f.close()` even when using `with` statements?
Or is it actually just completely safe to ignore this?

Reply:
Safe to ignore.  The `with` statement ensures that the file is closed upon exiting the `with` block, even in case of errors, so the `f.close()` will be trying to close an already-closed file.